balloc/balloc.c

```diff
@@ -2532,8 +2582,12 @@ static int allocate_blocks(int cr, struct balloc_allocation_context *bac,
 	if (rc == 0 && bac->bac_status == M0_BALLOC_AC_FOUND) {
 		if (len < m0_ext_length(&bac->bac_best))
 			balloc_new_preallocation(bac);
-		M0_ASSERT(is_extent_free(grp, &bac->bac_final, alloc_type,
-					 &cur));
+		if (!is_extent_free(grp, &bac->bac_final, alloc_type, &cur)) {
+			M0_ASSERT(is_extent_partof_best_extent(grp,
+				 &bac->bac_best, alloc_type, &cur));
+			M0_ASSERT(reset_final_extent(bac, cur, alloc_type));
+		}
```

Comment:
According to `doc/coding-style`:
>  * use assertions freely to verify state invariants. An asserted
       expression should have no side-effects;

The original code and the patch are both wrong.
These functions are modifying `cur`, and `the reset_final_extent()` function is modifying the 'final' and 'best'.

Reply:
The `Root Cause` of this problem is not fixed.
The "&bac->bac_final" should be in the free list. We have tried to find a suitable extent from free space. But the result is not in the free list.
This means, something is wrong in the "balloc_simple_scan_group()" or balloc_wild_scan_group()".

The "&bac->bac_best" should be the best candidate from the free space list.
The "&bac->bac_best" is used in the search. It is not used as a result.
So, no need to modify it here.

Reply:
can we assert by taking return code, do you want me to remove this assert ? previously M0_ASSERT was there for is_extent_free(&cur), so added further assert if is_extent_free() return false.

Reply:
balloc_simple_scan_group() looks ok , problem seems with balloc_wild_scan_group() because somehow it gives incorrect best
extent, best extent should be lies in free list of extents.

Reply:
Yes. The correct way is to take the return code, and assert the return code.